File: core/verifier.py
"""Generic temporal verifier — one engine for every sign.

verify(buffer, sign) returns a per-parameter breakdown PLUS an overall pass/fail. The overall
pass requires EVERY parameter marked required to individually clear its own threshold. There is
no averaging: a perfect handshape can never compensate for absent required movement.

Role assignment: the hand that moved more across the window is DOMINANT; the stiller one is
NONDOMINANT. This is handedness-agnostic and works for signs like COFFEE (one hand acts, one
holds) and HELP (both rise, dominant tracks the higher-motion A-hand on top).
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from core import handshape as hs
from core import movement as mv
from core import orientation as ori
from core.landmarks import RollingBuffer, normalized_distance
from core.schema import DOMINANT, NONDOMINANT, Anchor, MovementKind, Sign

SMOOTH_SECONDS = 0.5

# Anchor.CHEST geometry, in shoulder-widths below the shoulder line.
# The hand must be at chest HEIGHT: full credit within +-CHEST_VBAND of CHEST_OFFSET_RATIO,
# then a linear falloff over CHEST_VFALL. Tuned so the pass/fail boundary lands at ~dy 0.6:
# chest reads dy <= 0.6 (full credit up to 0.60), belly reads dy > 0.6 and fails.
CHEST_OFFSET_RATIO = 0.35
CHEST_VBAND = 0.25
CHEST_VFALL = 0.12

# Anchor.CHIN, measured against the REAL mouth landmark (robust to camera/proportions).
# When the fingertips touch the chin, the PALM CENTER (what we track) sits ~CHIN_DY shoulder-
# widths BELOW the mouth (the palm hangs below the fingertips). Full credit within +-CHIN_DY_BAND,
# linear falloff over CHIN_DY_FALL. This vertical band uniquely picks the chin and rejects the
# nose/face (palm ~at the mouth), the head (palm above the mouth), and the chest (palm too low).
CHIN_DY = 0.45
CHIN_DY_BAND = 0.18
CHIN_DY_FALL = 0.17


@dataclass
class ParamScore:
    name: str
    score: float
    threshold: float
    required: bool

    @property
    def cleared(self) -> bool:
        return self.score >= self.threshold

    @property
    def passed(self) -> bool:
        return (not self.required) or self.cleared


@dataclass
class VerifyResult:
    sign_name: str
    params: list[ParamScore]
    roles: dict

    @property
    def passed(self) -> bool:
        required = [p for p in self.params if p.required]
        return len(required) > 0 and all(p.passed for p in required)

    @property
    def failing_required(self) -> list[str]:
        return [p.name for p in self.params if p.required and not p.cleared]

    def get(self, name: str) -> ParamScore | None:
        for p in self.params:
            if p.name == name:
                return p
        return None


# ------------------------------------------------------------------ trajectory helpers
def _trajectory(buffer: RollingBuffer, handedness: str | None):
    if handedness is None:
        return []
    out = []
    for f in buffer:
        h = f.hand(handedness)
        if h is not None:
            out.append((f.t, h.center))
    return out


def _aligned_pair(buffer: RollingBuffer, label_a: str, label_b: str):
    """Frame-aligned (t, center) pairs for two hands (only frames where both are present)."""
    traj_a, traj_b = [], []
    for f in buffer:
        ha, hb = f.hand(label_a), f.hand(label_b)
        if ha is not None and hb is not None:
            traj_a.append((f.t, ha.center))
            traj_b.append((f.t, hb.center))
    return traj_a, traj_b


def _path_length(traj) -> float:
    if len(traj) < 2:
        return 0.0
    pts = np.array([c for _, c in traj], dtype=float)
    return float(np.sum(np.linalg.norm(np.diff(pts, axis=0), axis=1)))


def assign_roles(buffer: RollingBuffer) -> dict:
    """Map DOMINANT/NONDOMINANT to detected handedness labels by relative motion."""
    labels: list[str] = []
    for f in buffer:
        for h in f.hands:
            if h.handedness not in labels:
                labels.append(h.handedness)
    if not labels:
        return {}
    if len(labels) == 1:
        return {DOMINANT: labels[0]}
    labels.sort(key=lambda lab: _path_length(_trajectory(buffer, lab)), reverse=True)
    return {DOMINANT: labels[0], NONDOMINANT: labels[1]}


def _recent(buffer: RollingBuffer, seconds: float):
    frames = list(buffer)
    if not frames:
        return []
    end_t = frames[-1].t
    return [f for f in frames if end_t - f.t <= seconds]


def _latest_shoulder_width(buffer: RollingBuffer):
    for f in reversed(list(buffer)):
        if f.shoulder_width:
            return f.shoulder_width
    return None


# ------------------------------------------------------------------ parameter scorers
def _score_handshape(buffer, handedness, kind) -> float:
    vals = []
    for f in _recent(buffer, SMOOTH_SECONDS):
        h = f.hand(handedness)
        if h is not None:
            vals.append(hs.handshape_confidence(h, kind))
    return float(np.median(vals)) if vals else 0.0


def _best_fit_roles(buffer, sign: Sign, roles: dict) -> dict:
    """Stabilize dominant/nondominant for two-handed signs with DIFFERENT handshapes.

    assign_roles() picks the dominant hand by which one moved more — perfect for signs where one
    hand acts and one holds (COFFEE), but unstable for signs where BOTH hands move together (HELP:
    fist + open palm both lift). There the role labels flicker frame-to-frame, so the two handshape
    scores keep swapping 1.0<->0.0 and never line up. When the two declared handshapes differ, we
    instead keep whichever role assignment best FITS the declared shapes. Symmetric signs (COFFEE =
    fist+fist) and one-handed signs are left exactly as assign_roles returned them.
    """
    if not sign.two_handed or sign.nondominant is None:
        return roles
    if sign.dominant.kind == sign.nondominant.kind:
        return roles
    dl, nl = roles.get(DOMINANT), roles.get(NONDOMINANT)
    if dl is None or nl is None:
        return roles
    dk, nk = sign.dominant.kind, sign.nondominant.kind
    current = min(_score_handshape(buffer, dl, dk), _score_handshape(buffer, nl, nk))
    swapped = min(_score_handshape(buffer, nl, dk), _score_handshape(buffer, dl, nk))
    if swapped > current:
        return {DOMINANT: nl, NONDOMINANT: dl}
    return roles


def _score_location(buffer, sign: Sign, roles, shoulder_width) -> float:
    if shoulder_width is None:
        return 0.0
    loc = sign.location
    acting_label = roles.get(loc.acting_hand)
    if acting_label is None:
        return 0.0

    if loc.anchor == Anchor.CHIN:
        return _score_chin_reach(buffer, acting_label, shoulder_width, loc)

    vals = []
    for f in _recent(buffer, SMOOTH_SECONDS):
        acting = f.hand(acting_label)
        if acting is None:
            continue

        if loc.anchor == Anchor.OTHER_HAND:
            other_role = NONDOMINANT if loc.acting_hand == DOMINANT else DOMINANT
            other_label = roles.get(other_role)
            other = f.hand(other_label) if other_label else None
            if other is None:
                continue
            d = normalized_distance(acting.center, other.center, shoulder_width)
            dist_score = _band_score(d, loc.min_dist_ratio, loc.max_dist_ratio)
            vert_score = _vertical_score(loc.vertical, acting.center, other.center, shoulder_width)
            vals.append(min(dist_score, vert_score))
        elif loc.anchor == Anchor.CHEST:
            if f.left_shoulder is None or f.right_shoulder is None:
                continue
            mid = (f.left_shoulder + f.right_shoulder) / 2.0
            dx = abs(acting.center[0] - mid[0]) / shoulder_width      # sideways from center
            dy = (acting.center[1] - mid[1]) / shoulder_width         # below the shoulder line
            v = 1.0 - max(0.0, abs(dy - CHEST_OFFSET_RATIO) - CHEST_VBAND) / CHEST_VFALL
            h = 1.0 - max(0.0, dx - loc.max_dist_ratio) / 0.35
            vals.append(float(np.clip(min(v, h), 0.0, 1.0)))

        else:  # NEUTRAL_SPACE — anywhere in front of the torso, below the shoulders (loose)
            if f.left_shoulder is None or f.right_shoulder is None:
                continue
            mid = (f.left_shoulder + f.right_shoulder) / 2.0
            d = normalized_distance(acting.center, mid, shoulder_width)
            dist_score = _band_score(d, 0.0, loc.max_dist_ratio)
            below = acting.center[1] > mid[1]
            vals.append(dist_score if below else dist_score * 0.5)

    return float(np.median(vals)) if vals else 0.0


def _band_score(d: float, lo: float, hi: float) -> float:
    """1.0 inside [lo, hi]; falls off smoothly outside."""
    if lo <= d <= hi:
        return 1.0
    span = max(hi - lo, hi, 1e-6)
    if d < lo:
        return float(np.clip(1.0 - (lo - d) / span, 0.0, 1.0))
    return float(np.clip(1.0 - (d - hi) / span, 0.0, 1.0))


def _vertical_score(vertical, acting_c, other_c, shoulder_width) -> float:
    """Graded vertical preference, normalized to shoulder width.

    For "above": full credit when the acting hand is level-or-above the anchor; only penalized
    when it drops CLEARLY below (a soft ramp to 0 over ~1/3 shoulder width). This avoids the old
    binary flip that snapped location to 0 every time a grinding hand dipped to the same height.
    """
    if vertical is None:
        return 1.0
    # dy > 0 when the acting hand is ABOVE the anchor (image y grows downward)
    dy = (other_c[1] - acting_c[1]) / max(shoulder_width, 1e-6)
    ramp = 0.33  # how far below (in shoulder-widths) drives the score to 0
    if vertical == "above":
        return 1.0 if dy >= 0 else float(np.clip(1.0 + dy / ramp, 0.0, 1.0))
    if vertical == "below":
        return 1.0 if dy <= 0 else float(np.clip(1.0 - dy / ramp, 0.0, 1.0))
    return 1.0


def _score_chin_reach(buffer, acting_label, shoulder_width, loc) -> float:
    """Score whether the acting hand reached the chin, against the REAL mouth landmark.

    Robust to camera angle/distance and proportions: we compare the hand to the detected mouth,
    not a guessed offset. Take the hand's CLOSEST approach to the mouth among frames where the
    hand is at-or-below mouth height (frames well ABOVE the mouth are forehead/head, not chin, and
    are excluded). A hand that stays down at the chest never gets near the mouth, so it scores ~0.
    """
    best = 0.0
    for f in buffer:
        h = f.hand(acting_label)
        if h is None or f.mouth is None:
            continue
        dx = abs(h.center[0] - f.mouth[0]) / shoulder_width
        dy = (h.center[1] - f.mouth[1]) / shoulder_width    # +ve = palm below the mouth (the chin)
        vert = 1.0 - max(0.0, abs(dy - CHIN_DY) - CHIN_DY_BAND) / CHIN_DY_FALL
        horiz = 1.0 - max(0.0, dx - loc.max_dist_ratio) / 0.35
        best = max(best, min(vert, horiz))                  # best (closest) approach over the window
    return float(np.clip(best, 0.0, 1.0))


def _score_movement(buffer, sign: Sign, roles, shoulder_width) -> float:
    req = sign.movement
    if req.kind == MovementKind.NONE:
        return 1.0
    actor_label = roles.get(req.actor)
    actor_traj = _trajectory(buffer, actor_label)
    if shoulder_width is None or not actor_traj:
        return 0.0
    if req.kind == MovementKind.CONVERGE:
        # Two-hand movement: need aligned trajectories for both hands.
        ndom_label = roles.get(NONDOMINANT)
        if ndom_label is None:
            return 0.0
        traj_a, traj_b = _aligned_pair(buffer, actor_label, ndom_label)
        return mv.converge_confidence(traj_a, traj_b, shoulder_width, req)
    return mv.movement_confidence(actor_traj, shoulder_width, req)


def _score_orientation(buffer, sign: Sign, roles) -> float:
    o = sign.orientation
    label = roles.get(o.hand)
    if label is None:
        return 0.0
    vals = []
    for f in _recent(buffer, SMOOTH_SECONDS):
        h = f.hand(label)
        if h is not None:
            vals.append(ori.facing_confidence(h, o.facing))
    return float(np.median(vals)) if vals else 0.0


# ------------------------------------------------------------------ public API
def movement_debug(buffer: RollingBuffer, sign: Sign) -> str:
    """One-line readout of live movement sub-metrics for the dev demo / calibration."""
    req = sign.movement
    if req.kind == MovementKind.NONE:
        return "static (no movement required)"
    roles = assign_roles(buffer)
    sw = _latest_shoulder_width(buffer)
    traj = _trajectory(buffer, roles.get(req.actor))
    if req.kind == MovementKind.CIRCULAR:
        m = mv.circular_metrics(traj, sw if sw else 0.0, req)
        return (f"rot {m.net_rotation_deg:3.0f}/{req.min_total_rotation_deg:.0f}deg  "
                f"radCV {m.radius_cv:0.2f}(full<0.30)  r/sw {m.mean_r_ratio:0.2f}  "
                f"frames {m.n}  {m.duration:0.1f}s")
    if req.kind == MovementKind.CONVERGE:
        ndom_label = roles.get(NONDOMINANT)
        if ndom_label and sw:
            traj_a, traj_b = _aligned_pair(buffer, roles.get(req.actor), ndom_label)
            n = min(len(traj_a), len(traj_b))
            if n >= 2:
                pts_a = np.array([c for _, c in traj_a[:n]], float)
                pts_b = np.array([c for _, c in traj_b[:n]], float)
                gap = np.linalg.norm(pts_a - pts_b, axis=1) / sw
                return f"converge: gap {gap[0]:.2f}->{gap[-1]:.2f}sw  n={n}"
        return "converge: waiting for both hands"
    return f"{req.kind.value}: {len(traj)} samples"


def location_debug(buffer: RollingBuffer, sign: Sign) -> str:
    """One-line readout of the acting hand's position vs the shoulders, for calibration.

    dy = shoulder-widths BELOW the shoulder line, dx = shoulder-widths SIDEWAYS from center.
    For a CHEST sign it also shows the target band so you can see where chest ends / belly begins.
    """
    loc = sign.location
    roles = assign_roles(buffer)
    sw = _latest_shoulder_width(buffer)
    label = roles.get(loc.acting_hand)
    for f in reversed(_recent(buffer, SMOOTH_SECONDS)):
        h = f.hand(label) if label else None
        if h is not None and f.left_shoulder is not None and sw:
            mid = (f.left_shoulder + f.right_shoulder) / 2.0
            dx = abs(h.center[0] - mid[0]) / sw
            dy = (h.center[1] - mid[1]) / sw
            if loc.anchor == Anchor.CHEST:
                lo, hi = CHEST_OFFSET_RATIO - CHEST_VBAND, CHEST_OFFSET_RATIO + CHEST_VBAND
                return f"loc dy {dy:.2f} (chest band {lo:.2f}-{hi:.2f})  dx {dx:.2f}"
            if loc.anchor == Anchor.CHIN:
                dys = [(g.hand(label).center[1] - g.mouth[1]) / sw
                       for g in buffer if g.hand(label) is not None and g.mouth is not None]
                best_dy = min(dys, key=lambda v: abs(v - CHIN_DY)) if dys else dy
                return f"loc palm-vs-mouth dy {best_dy:+.2f} (chin ~ +{CHIN_DY:.2f})  now dy {dy:+.2f}"
            return f"loc dy {dy:.2f}  dx {dx:.2f}  anchor={loc.anchor.value}"
    return "loc: (acting hand or shoulders not visible)"


def verify(buffer: RollingBuffer, sign: Sign) -> VerifyResult:
    roles = _best_fit_roles(buffer, sign, assign_roles(buffer))
    sw = _latest_shoulder_width(buffer)
    params: list[ParamScore] = []

    dom = sign.dominant
    params.append(ParamScore(
        "handshape_dominant",
        _score_handshape(buffer, roles.get(DOMINANT), dom.kind),
        dom.min_confidence, dom.required,
    ))

    if sign.nondominant is not None:
        nd = sign.nondominant
        params.append(ParamScore(
            "handshape_nondominant",
            _score_handshape(buffer, roles.get(NONDOMINANT), nd.kind),
            nd.min_confidence, nd.required,
        ))

    params.append(ParamScore(
        "location",
        _score_location(buffer, sign, roles, sw),
        sign.location.min_confidence, sign.location.required,
    ))

    params.append(ParamScore(
        "movement",
        _score_movement(buffer, sign, roles, sw),
        sign.movement.min_confidence, sign.movement.required,
    ))

    if sign.orientation is not None:
        params.append(ParamScore(
            "orientation",
            _score_orientation(buffer, sign, roles),
            sign.orientation.min_confidence, sign.orientation.required,
        ))

    return VerifyResult(sign.name, params, roles)

File: web/src/engine/verifier.ts
import { clip, median } from './math-utils';
import { handshapeConfidence } from './handshape';
import * as mv from './movement';
import { facingConfidence } from './orientation';
import {
  type Frame, type RollingBuffer,
  handCenter, frameHand, frameShoulderWidth, normalizedDistance,
} from './landmarks';
import {
  DOMINANT, NONDOMINANT,
  Anchor, MovementKind,
  type Sign,
} from './schema';

const SMOOTH_SECONDS = 0.5;

const CHEST_OFFSET_RATIO = 0.35;
const CHEST_VBAND = 0.25;
const CHEST_VFALL = 0.12;

const CHIN_DY = 0.45;
const CHIN_DY_BAND = 0.18;
const CHIN_DY_FALL = 0.17;

const FOREHEAD_DY_MAX = 0.15;
const FOREHEAD_DY_FALL = 0.30;

const BELLY_DY = 0.90;
const BELLY_DY_BAND = 0.30;
const BELLY_DY_FALL = 0.25;

const SHOULDER_FALL = 0.30;

export interface ParamScore {
  name: string;
  score: number;
  threshold: number;
  required: boolean;
}

export function paramCleared(p: ParamScore): boolean {
  return p.score >= p.threshold;
}

export function paramPassed(p: ParamScore): boolean {
  return !p.required || paramCleared(p);
}

export interface VerifyResult {
  signName: string;
  params: ParamScore[];
  roles: Record<string, string>;
}

export function resultPassed(r: VerifyResult): boolean {
  const required = r.params.filter((p) => p.required);
  return required.length > 0 && required.every(paramPassed);
}

export function resultFailingRequired(r: VerifyResult): string[] {
  return r.params.filter((p) => p.required && !paramCleared(p)).map((p) => p.name);
}

export function resultGet(r: VerifyResult, name: string): ParamScore | undefined {
  return r.params.find((p) => p.name === name);
}

type Traj = [number, number[]][];

function trajectory(buffer: RollingBuffer, handedness: string | null): Traj {
  if (!handedness) return [];
  const out: Traj = [];
  for (const f of buffer) {
    const h = frameHand(f, handedness);
    if (h) out.push([f.t, handCenter(h)]);
  }
  return out;
}

function alignedPair(
  buffer: RollingBuffer,
  labelA: string,
  labelB: string
): [Traj, Traj] {
  const trajA: Traj = [];
  const trajB: Traj = [];
  for (const f of buffer) {
    const ha = frameHand(f, labelA);
    const hb = frameHand(f, labelB);
    if (ha && hb) {
      trajA.push([f.t, handCenter(ha)]);
      trajB.push([f.t, handCenter(hb)]);
    }
  }
  return [trajA, trajB];
}

function pathLength(traj: Traj): number {
  if (traj.length < 2) return 0;
  let total = 0;
  for (let i = 1; i < traj.length; i++) {
    const dx = traj[i][1][0] - traj[i - 1][1][0];
    const dy = traj[i][1][1] - traj[i - 1][1][1];
    total += Math.sqrt(dx * dx + dy * dy);
  }
  return total;
}

export function assignRoles(buffer: RollingBuffer): Record<string, string> {
  const labels: string[] = [];
  for (const f of buffer) {
    for (const h of f.hands) {
      if (!labels.includes(h.handedness)) labels.push(h.handedness);
    }
  }
  if (labels.length === 0) return {};
  if (labels.length === 1) return { [DOMINANT]: labels[0] };
  labels.sort((a, b) => pathLength(trajectory(buffer, b)) - pathLength(trajectory(buffer, a)));
  return { [DOMINANT]: labels[0], [NONDOMINANT]: labels[1] };
}

function recent(buffer: RollingBuffer, seconds: number): Frame[] {
  const frames = buffer.frames;
  if (frames.length === 0) return [];
  const endT = frames[frames.length - 1].t;
  return frames.filter((f) => endT - f.t <= seconds);
}

function latestShoulderWidth(buffer: RollingBuffer): number | null {
  const frames = buffer.frames;
  for (let i = frames.length - 1; i >= 0; i--) {
    const sw = frameShoulderWidth(frames[i]);
    if (sw) return sw;
  }
  return null;
}

function scoreHandshape(buffer: RollingBuffer, handedness: string | null, kind: string): number {
  if (!handedness) return 0;
  const vals: number[] = [];
  for (const f of recent(buffer, SMOOTH_SECONDS)) {
    const h = frameHand(f, handedness);
    if (h) vals.push(handshapeConfidence(h, kind));
  }
  return vals.length > 0 ? median(vals) : 0;
}

function bestFitRoles(
  buffer: RollingBuffer,
  sign: Sign,
  roles: Record<string, string>
): Record<string, string> {
  if (!sign.twoHanded || !sign.nondominant) return roles;
  if (sign.dominant.kind === sign.nondominant.kind) return roles;
  const dl = roles[DOMINANT];
  const nl = roles[NONDOMINANT];
  if (!dl || !nl) return roles;
  const dk = sign.dominant.kind;
  const nk = sign.nondominant.kind;
  const current = Math.min(scoreHandshape(buffer, dl, dk), scoreHandshape(buffer, nl, nk));
  const swapped = Math.min(scoreHandshape(buffer, nl, dk), scoreHandshape(buffer, dl, nk));
  if (swapped > current) return { [DOMINANT]: nl, [NONDOMINANT]: dl };
  return roles;
}

function bandScore(d: number, lo: number, hi: number): number {
  if (lo <= d && d <= hi) return 1;
  const span = Math.max(hi - lo, hi, 1e-6);
  if (d < lo) return clip(1 - (lo - d) / span, 0, 1);
  return clip(1 - (d - hi) / span, 0, 1);
}

function verticalScore(
  vertical: 'above' | 'below' | null | undefined,
  actingC: number[],
  otherC: number[],
  shoulderWidth: number
): number {
  if (!vertical) return 1;
  const dy = (otherC[1] - actingC[1]) / Math.max(shoulderWidth, 1e-6);
  const ramp = 0.33;
  if (vertical === 'above') return dy >= 0 ? 1 : clip(1 + dy / ramp, 0, 1);
  if (vertical === 'below') return dy <= 0 ? 1 : clip(1 - dy / ramp, 0, 1);
  return 1;
}

function scoreChinReach(
  buffer: RollingBuffer,
  actingLabel: string,
  shoulderWidth: number,
  loc: Sign['location']
): number {
  let best = 0;
  for (const f of buffer) {
    const h = frameHand(f, actingLabel);
    if (!h || !f.mouth) continue;
    const c = handCenter(h);
    const dx = Math.abs(c[0] - f.mouth[0]) / shoulderWidth;
    const dy = (c[1] - f.mouth[1]) / shoulderWidth;
    const vert = 1 - Math.max(0, Math.abs(dy - CHIN_DY) - CHIN_DY_BAND) / CHIN_DY_FALL;
    const horiz = 1 - Math.max(0, dx - loc.maxDistRatio) / 0.35;
    best = Math.max(best, Math.min(vert, horiz));
  }
  return clip(best, 0, 1);
}

function scoreLocation(
  buffer: RollingBuffer,
  sign: Sign,
  roles: Record<string, string>,
  shoulderWidth: number | null
): number {
  if (!shoulderWidth) return 0;
  const loc = sign.location;
  const actingLabel = roles[loc.actingHand];
  if (!actingLabel) return 0;

  if (loc.anchor === Anchor.CHIN) {
    return scoreChinReach(buffer, actingLabel, shoulderWidth, loc);
  }

  const vals: number[] = [];
  for (const f of recent(buffer, SMOOTH_SECONDS)) {
    const acting = frameHand(f, actingLabel);
    if (!acting) continue;
    const ac = handCenter(acting);

    if (loc.anchor === Anchor.OTHER_HAND) {
      const otherRole = loc.actingHand === DOMINANT ? NONDOMINANT : DOMINANT;
      const otherLabel = roles[otherRole];
      const other = otherLabel ? frameHand(f, otherLabel) : null;
      if (!other) continue;
      const oc = handCenter(other);
      let d: number;
      if (loc.useClosestApproach) {
        let minD = Infinity;
        for (const ap of acting.points) {
          for (const op of other.points) {
            const dx = ap[0] - op[0], dy = ap[1] - op[1];
            minD = Math.min(minD, Math.sqrt(dx * dx + dy * dy));
          }
        }
        d = minD / shoulderWidth;
      } else {
        d = normalizedDistance(ac, oc, shoulderWidth);
      }
      const distScore = bandScore(d, loc.minDistRatio, loc.maxDistRatio);
      const vScore = verticalScore(loc.vertical, ac, oc, shoulderWidth);
      let score = Math.min(distScore, vScore);
      if (loc.below === 'mouth' && f.mouth) {
        const dy = (ac[1] - f.mouth[1]) / shoulderWidth;
        score *= clip((dy + 0.1) / 0.3, 0, 1);
      }
      vals.push(score);
    } else if (loc.anchor === Anchor.CHEST) {
      if (!f.leftShoulder || !f.rightShoulder) continue;
      const mid = [(f.leftShoulder[0] + f.rightShoulder[0]) / 2, (f.leftShoulder[1] + f.rightShoulder[1]) / 2];
      const dx = Math.abs(ac[0] - mid[0]) / shoulderWidth;
      const dy = (ac[1] - mid[1]) / shoulderWidth;
      const v = 1 - Math.max(0, Math.abs(dy - CHEST_OFFSET_RATIO) - CHEST_VBAND) / CHEST_VFALL;
      const h = 1 - Math.max(0, dx - loc.maxDistRatio) / 0.35;
      vals.push(clip(Math.min(v, h), 0, 1));
    } else if (loc.anchor === Anchor.FOREHEAD) {
      if (!f.mouth) continue;
      const dx = Math.abs(ac[0] - f.mouth[0]) / shoulderWidth;
      const dy = (ac[1] - f.mouth[1]) / shoulderWidth;
      const v = 1 - Math.max(0, dy - FOREHEAD_DY_MAX) / FOREHEAD_DY_FALL;
      const h = 1 - Math.max(0, dx - loc.maxDistRatio) / 0.4;
      vals.push(clip(Math.min(v, h), 0, 1));
    } else if (loc.anchor === Anchor.BELLY) {
      if (!f.leftShoulder || !f.rightShoulder) continue;
      const mid = [(f.leftShoulder[0] + f.rightShoulder[0]) / 2, (f.leftShoulder[1] + f.rightShoulder[1]) / 2];
      const dx = Math.abs(ac[0] - mid[0]) / shoulderWidth;
      const dy = (ac[1] - mid[1]) / shoulderWidth;
      const v = 1 - Math.max(0, Math.abs(dy - BELLY_DY) - BELLY_DY_BAND) / BELLY_DY_FALL;
      const h = 1 - Math.max(0, dx - loc.maxDistRatio) / 0.4;
      vals.push(clip(Math.min(v, h), 0, 1));
    } else if (loc.anchor === Anchor.SHOULDER) {
      if (!f.leftShoulder || !f.rightShoulder) continue;
      const dL = normalizedDistance(ac, f.leftShoulder, shoulderWidth);
      const dR = normalizedDistance(ac, f.rightShoulder, shoulderWidth);
      const d = Math.min(dL, dR);
      vals.push(clip(1 - Math.max(0, d - loc.maxDistRatio) / SHOULDER_FALL, 0, 1));
    } else {
      // NEUTRAL_SPACE
      if (!f.leftShoulder || !f.rightShoulder) continue;
      const mid = [(f.leftShoulder[0] + f.rightShoulder[0]) / 2, (f.leftShoulder[1] + f.rightShoulder[1]) / 2];
      const d = normalizedDistance(ac, mid, shoulderWidth);
      const distScore = bandScore(d, 0, loc.maxDistRatio);
      const below = ac[1] > mid[1];
      vals.push(below ? distScore : distScore * 0.5);
    }
  }

  return vals.length > 0 ? median(vals) : 0;
}

function scoreMovement(
  buffer: RollingBuffer,
  sign: Sign,
  roles: Record<string, string>,
  shoulderWidth: number | null
): number {
  const req = sign.movement;
  if (req.kind === MovementKind.NONE) return 1;
  const actorLabel = roles[req.actor];
  const actorTraj = trajectory(buffer, actorLabel ?? null);
  if (!shoulderWidth || actorTraj.length === 0) return 0;

  if (req.kind === MovementKind.CONVERGE) {
    const ndomLabel = roles[NONDOMINANT];
    if (!ndomLabel) return 0;
    const [trajA, trajB] = alignedPair(buffer, actorLabel!, ndomLabel);
    return mv.convergeConfidence(trajA, trajB, shoulderWidth, req);
  }
  return mv.movementConfidence(actorTraj, shoulderWidth, req);
}

function scoreOrientation(
  buffer: RollingBuffer,
  sign: Sign,
  roles: Record<string, string>
): number {
  const o = sign.orientation;
  if (!o) return 0;
  const label = roles[o.hand];
  if (!label) return 0;
  const vals: number[] = [];
  for (const f of recent(buffer, SMOOTH_SECONDS)) {
    const h = frameHand(f, label);
    if (h) vals.push(facingConfidence(h, o.facing));
  }
  return vals.length > 0 ? median(vals) : 0;
}

export function verify(buffer: RollingBuffer, sign: Sign): VerifyResult {
  const roles = bestFitRoles(buffer, sign, assignRoles(buffer));
  const sw = latestShoulderWidth(buffer);
  const params: ParamScore[] = [];

  const dom = sign.dominant;
  params.push({
    name: 'handshape_dominant',
    score: scoreHandshape(buffer, roles[DOMINANT] ?? null, dom.kind),
    threshold: dom.minConfidence,
    required: dom.required,
  });

  if (sign.nondominant) {
    const nd = sign.nondominant;
    params.push({
      name: 'handshape_nondominant',
      score: scoreHandshape(buffer, roles[NONDOMINANT] ?? null, nd.kind),
      threshold: nd.minConfidence,
      required: nd.required,
    });
  }

  params.push({
    name: 'location',
    score: scoreLocation(buffer, sign, roles, sw),
    threshold: sign.location.minConfidence,
    required: sign.location.required,
  });

  params.push({
    name: 'movement',
    score: scoreMovement(buffer, sign, roles, sw),
    threshold: sign.movement.minConfidence,
    required: sign.movement.required,
  });

  if (sign.orientation) {
    params.push({
      name: 'orientation',
      score: scoreOrientation(buffer, sign, roles),
      threshold: sign.orientation.minConfidence,
      required: sign.orientation.required,
    });
  }

  return { signName: sign.name, params, roles };
}
